charts/CSVGenerator.tsx

```diff
@@ -108,18 +108,24 @@ export class CSVGenerator {
         dim: ChartDimensionWithOwidVariable,
         entity: string
     ): [string, string | number] | null {
-        let latestYear = undefined
-        let latestValue = undefined
-
-        const latestYearValue = dim.yearAndValueOfLatestValueforEntity(entity)
-
-        if (latestYearValue) {
-            latestYear = dim.formatYear(first(latestYearValue) as number)
-            latestValue = last(latestYearValue)
+        let year = undefined
+        let value = undefined
+
+        if (dim.targetYear && dim.tolerance) {
+            const dataValues = valuesByEntityAtYears(
+                dim.valueByEntityAndYear,
+                [dim.targetYear],
+                dim.tolerance
+            ).get(entity)
```

Comment:
This computation is repeated for each entity, which isn't great. But it also seems like it might be some work to try to refactor, and the code wouldn't be as elegant, so I think it's fine to leave it as it is.

Potentially a DataFrame or similar abstraction will make this more efficient and simpler in the future – having the concept of a "join" is exactly what we need between the "single value" and "all values" data frames.

Reply:
Actually, I pushed a branch ([`csv_uses_x_axis_override_optimization`](https://github.com/owid/owid-grapher/commit/1c0cd51a9521533dfa94e739bbaf0ccfdfa69f79)) that extracts a `valuesAtYears()` utility function:

https://github.com/owid/owid-grapher/blob/f1b92b9cd9158f0bda4a0f81578b71dbcca9618b/charts/Util.ts#L591-L609

Reply:
Ah, that was a heavy function! Very good catch, merged it in.